docs/docs/modules/data_connection/retrievers/custom_retriever.ipynb

```diff
@@ -0,0 +1,309 @@
+{
+ "cells": [
+  {
+   "cell_type": "raw",
+   "id": "b5fc1fc7-c4c5-418f-99da-006c604a7ea6",
+   "metadata": {},
+   "source": [
+    "---\n",
+    "title: Custom Retriever\n",
+    "---"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "ff6f3c79-0848-4956-9115-54f6b2134587",
+   "metadata": {},
+   "source": [
+    "# Custom Retriever\n",
+    "\n",
+    "## Overview\n",
+    "\n",
+    "Many LLM applications involve retrieving information from external data sources using a `Retriever`. \n",
+    "\n",
+    "A retriever is responsible for retrieving a list of relevant `Documents` to a given user `query`.\n",
+    "\n",
+    "The retrieved documents are often formatted into prompts that are fed into an LLM, allowing the LLM to use the information in the to generate an appropriate response (e.g., answering a user question based on a knowledge base).\n",
+    "\n",
+    "## Interface\n",
+    "\n",
+    "To create your own retriever, you need to extend the `BaseRetriever` class and implement the following methods:\n",
+    "\n",
+    "| Method                         | Description                                      | Required/Optional |\n",
+    "|--------------------------------|--------------------------------------------------|-------------------|\n",
+    "| `_get_relevant_documents`      | Get documents relevant to a query.               | Required          |\n",
+    "| `_aget_relevant_documents`     | Implement to provide async native support.       | Optional          |\n",
+    "\n",
+    "\n",
+    "The logic inside of `_get_relevant_documents` can involve arbitrary calls to a database or to the web using requests.\n",
+    "\n",
+    ":::{.callout-tip}\n",
+    "By inherting from `BaseRetriever`, your retriever automatically becomes a `LangChain` Runnable and will gain the standard `Runnable` functionality out of the box!\n",
+    ":::\n",
+    "\n",
+    "\n",
+    ":::{.callout-info}\n",
+    "You can use a `RunnableLambda` or `RunnableGenerator` to implement a retriever.\n",
+    "\n",
+    "The main benefit of implementing a retriever as a `BaseRetriever` vs. a `RunnableLambda` is that a `BaseRetriever` is a well\n",
```

Comment:
```suggestion
    "The main benefit of implementing a retriever as a `BaseRetriever` vs. a `RunnableLambda` (a custom [runnable function](/docs/expression_language/primitives/functions)) is that a `BaseRetriever` is a well\n",
```

Reply:
Not sure this will render in a callout. Going to skip for now. I also think that we need a better strategy for this with references automatically inserted